src/lib.rs

```diff
@@ -1,40 +1,97 @@
-extern crate byteorder;
-#[macro_use]
-extern crate enum_primitive;
-extern crate num;
+use std::any::type_name;
+use std::io::{Read, Seek};
 
-use std::fmt;
-use std::io::Result;
+use num_traits::FromPrimitive;
+use tiff::decoder::{Decoder, DecodingResult};
+use tiff::tags::Tag;
+use tiff::TiffResult;
 
-pub mod geotiff;
-mod lowlevel;
-mod reader;
+use crate::image_data::*;
 
-pub use geotiff::TIFF;
-use reader::*;
+mod image_data;
 
-/// The GeoTIFF library reads `.tiff` files.
+macro_rules! assert_type {
+    ($result: expr, $actual: ty, $expected: ty) => {
+        $result
+            .ok_or_else(|| {
+                format!(
+                    "Cannot represent {} as {}",
+                    type_name::<$actual>(),
+                    type_name::<$expected>()
+                )
+            })
+            .unwrap()
+    };
+}
+
+/// The basic GeoTIFF struct. This includes any metadata as well as the actual image data.
 ///
-/// It is primarily used within a routing application that needs to parse digital elevation models.
-/// As such, other use cases are NOT tested (for now).
-impl TIFF {
-    /// Opens a `.tiff` file at the location indicated by `filename`.
-    pub fn open(filename: &str) -> Result<Box<TIFF>> {
-        let tiff_reader = TIFFReader;
-        tiff_reader.load(filename)
-    }
+/// The image data has a size of width * length * num_samples
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
```

Comment:
Thoughts on separating this into `width` and `height`? I always get tripped up on whether it's x/y or y/x.

Reply:
I can do that, of course. I just used the format of the underlying `tiff` crate. We should maybe also make it clear that we are referring to raster coordinates as opposed to model coordinates, calling it  `rasterWidth` and `rasterHeight`.

Reply:
Yep, `rasterWidth` and `rasterHeight` sounds good to me!